user/mongoStoreClient.go

```diff
@@ -1,104 +1,171 @@
-	"github.com/globalsign/mgo/bson"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
 
-	"github.com/tidepool-org/go-common/clients/mongo"
+	tpMongo "github.com/tidepool-org/go-common/clients/mongo"
 )
 
 const (
-	USERS_COLLECTION  = "users"
-	TOKENS_COLLECTION = "tokens"
+	usersCollectionName  = "users"
+	tokensCollectionName = "tokens"
+	userStoreAPIPrefix   = "api/user/store "
 )
 
+// Because the `users` collection already exists on all environments (especially `prd`),
+// and MongoDB doesn't allow modification of default collation on an existing collection,
+// we need to specify collation manually everywhere we generate an index, or make a query
+// with the notable exception of the `_id` field
+var usersCollation *options.Collation = &options.Collation{Locale: "en", Strength: 1}
+
+// MongoStoreClient - Mongo Storage Client
 type MongoStoreClient struct {
-	session *mgo.Session
+	client   *mongo.Client
+	context  context.Context
+	database string
 }
 
-//We implement the interface from user.Storage
-func NewMongoStoreClient(config *mongo.Config) *MongoStoreClient {
+// NewMongoStoreClient creates a new MongoStoreClient
+func NewMongoStoreClient(config *tpMongo.Config) *MongoStoreClient {
+	connectionString, err := config.ToConnectionString()
+	if err != nil {
+		log.Fatal(userStoreAPIPrefix, fmt.Sprintf("Invalid MongoDB configuration: %s", err))
+	}
 
-	mongoSession, err := mongo.Connect(config)
+	clientOptions := options.Client().ApplyURI(connectionString)
+	mongoClient, err := mongo.Connect(context.Background(), clientOptions)
 	if err != nil {
-		log.Fatalf("Cannot connect to mongo: %v, %v", config, err)
+		log.Fatal(userStoreAPIPrefix, fmt.Sprintf("Invalid MongoDB connection string: %s", err))
 	}
 
 	return &MongoStoreClient{
-		session: mongoSession,
+		client:   mongoClient,
+		context:  context.Background(),
+		database: config.Database,
 	}
 }
 
-func mgoUsersCollection(cpy *mgo.Session) *mgo.Collection {
-	return cpy.DB("").C(USERS_COLLECTION)
+// WithContext returns a shallow copy of c with its context changed
+// to ctx. The provided ctx must be non-nil.
+func (msc *MongoStoreClient) WithContext(ctx context.Context) Storage {
+	if ctx == nil {
+		panic("nil context")
+	}
+	msc2 := new(MongoStoreClient)
+	*msc2 = *msc
+	msc2.context = ctx
+	return msc2
 }
 
-func mgoTokensCollection(cpy *mgo.Session) *mgo.Collection {
-	return cpy.DB("").C(TOKENS_COLLECTION)
-}
+// EnsureIndexes exist for the MongoDB collection. EnsureIndexes uses the Background() context, in order
+// to pass back the MongoDB errors, rather than any context errors.
+func (msc *MongoStoreClient) EnsureIndexes() error {
+	usersIndexes := []mongo.IndexModel{
+		{
+			Keys: bson.D{{Key: "userid", Value: 1}},
+			Options: options.Index().
+				SetCollation(usersCollation).
+				SetUnique(true).
+				SetBackground(true),
+		},
+		{
+			Keys: bson.D{{Key: "username", Value: 1}},
```

Comment:
Shouldn't `username` and `emails` be unique too?

Reply:
Ideally, but not for this PR. See https://tidepool.atlassian.net/browse/BACK-1376